library/alloc/src/vec/mod.rs

```diff
@@ -1987,6 +2019,49 @@ pub fn from_elem_in<T: Clone, A: Allocator>(elem: T, n: usize, alloc: A) -> Vec<
     <T as SpecFromElem>::from_elem(elem, n, alloc)
 }
 
+trait ExtendFromWithinSpec {
+    /// Safety: src need to be valid index
+    unsafe fn spec_extend_from_within(&mut self, src: Range<usize>);
+}
+
+impl<T: Clone> ExtendFromWithinSpec for Vec<T> {
+    default unsafe fn spec_extend_from_within(&mut self, src: Range<usize>) {
+        self.reserve(src.len());
+        let i = {
+            // copied from `spare_capacity_mut` b/c it doesn't allow access to the content
+            let spare_space = unsafe {
+                slice::from_raw_parts_mut(
+                    self.as_mut_ptr().add(self.len) as *mut MaybeUninit<T>,
+                    self.buf.capacity() - self.len,
+                )
+            };
```

Comment:
I think this example shows that we should have a method that splits `Vec` into used+spared slices. Something like this: 
```rust
// There probably is a better name... 
pub fn split_spare(&mut self) -> (&mut [T], &mut [MaybeUninit<T>]) { ... }
```
It's too restrictive to only be able to access either of spaces (used/spare) at a time.

Reply:
I sure can! I've thought of a public one, but it will probable be easier to make it pub in a different PR.